WIP: initial WebAssembly exception-handling support.

cfallin opened the pull request:
This PR introduces support for the [Wasm exception-handling proposal], which introduces a conventional try/catch mechanism to WebAssembly. The PR supports modules that use `try_table` to register handlers for a lexical scope; and provides `throw` and `throw_ref` that allocate (in the first case) and throw exception objects.

This PR builds on top of the work in #10510 for Cranelift-level exception support, #10919 for an unwinder, and #11230 for exception objects built on top of GC, in addition a bunch of smaller fix and enabling PRs around those. It is currently stacked on top of #11321.

This PR does not yet provide host-boundary-crossing exceptions; exceptions that are not caught in a given Wasm activation become traps at the host boundary. That support will come in a subsequent PR.

Because exceptions do not yet cross the host boundary, this also does not yet enable the `assert_exception` wast directive, and so cannot yet support the spec-tests. That will also come in a subsequent PR.

[Wasm exception-handling proposal]: https://github.com/WebAssembly/exception-handling/

cfallin commented:
Logistical note: I'm posting this as a draft now to get early feedback and because I know folks are waiting to see how it is shaping up. I'm on vacation for two weeks starting now (back Mon Aug 11) and will plan to polish then. I'm hoping to actually get host-boundary integration built as well, if I can, in this PR, to enable spec-tests, but if that turns out to be too much then it will come right after. Following that, fuzzing is the only piece that remains, I think.

fitzgen reviewed (commented):
Shaping up nicely!

We had talked elsewhere about removing the exception composite type variant and having tags refer to just their function type (but we would now optionally have a GC struct layout for a function type's parameters for use with exceptions). This would better align us with the Wasm spec and make it so that there are less new additions to the types registry code and also fewer interactions at runtime with its locking and tables and all that. Are you still planning on pursuing this?

cfallin commented:
> We had talked elsewhere about removing the exception composite type variant and having tags refer to just their function type (but we would now optionally have a GC struct layout for a function type's parameters for use with exceptions). This would better align us with the Wasm spec and make it so that there are less new additions to the types registry code and also fewer interactions at runtime with its locking and tables and all that. Are you still planning on pursuing this?

Ah, sorry, I hadn't made a note in the PR message here, but: I tried and abandoned that path. (Or more precisely, having exception layouts hang off of the function type; `TagType` is already a thin newtype wrapper around `FuncType`.) At a high level, pulling on that string seemed to unwind way too much structure and hit too many places that really still wanted a concrete type for an exception object. If you're curious, my WIP branch is [here](https://github.com/cfallin/wasmtime/tree/wasm-exception-types) (still has many type-errors, mid-refactor). In essence I think that path leads to more complexity, not less, unfortunately.

> also fewer interactions at runtime with its locking and tables and all that

The current implementation performs no locking or accesses to the type registry at runtime; it uses the dynamic context mechanism in Cranelift to get straight to the instance, then look up tags (`VMTagDefinition`s), and compares tag IDs (instance-id/defined-tag-index). In particular, the tag information is at a constant offset in the exception object (similar to array lengths) so we don't need a layout to write the generic path.

(Still on PTO but will respond occasionally to keep review moving)

cfallin commented on crates/wasmtime/src/runtime/type_registry.rs:
```diff
@@ -1336,6 +1336,23 @@ impl TypeRegistry {
+    /// Borrow the GC layout for teh given index's type.
+    /// Returns `None` for types that do not have GC layouts (i.e. function
+    /// types).
+    pub fn with_layout<R>(
```
Sorry, this is left over from an earlier implementation that fetched layouts dynamically before I had put the tag info at a constant offset in exception objects. (Or rather, it was always at a constant offset per allocator, but I hadn't done the plumbing that array-lengths have to avoid getting it from the layout.)

I'll remove this; but to clarify, we have no need to access layouts or the type registry at all at runtime with the exceptions implementation. The unwinder is completely generic over exception layouts and only compares tag identities.

cfallin commented on crates/wasmtime/src/runtime/module.rs:
```diff
@@ -1126,6 +1128,13 @@ impl Module {
+    /// Obtain an exception-table parser on this module's exception metadata.
+    #[cfg(feature = "gc")]
+    pub(crate) fn exception_table<'a>(&'a self) -> ExceptionTable<'a> {
+        ExceptionTable::parse(self.inner.code.code_memory().exception_tables())
```
Yes -- it reads two `u32`s for lengths of arrays, then builds an `ExceptionTable` with slices over arrays.

cfallin commented on crates/wasmtime/src/config.rs:
```diff
@@ -2212,6 +2217,13 @@ impl Config {
 
+        #[cfg(feature = "gc")]
+        if self.enabled_features.contains(WasmFeatures::EXCEPTIONS)
+            && self.disabled_features.contains(WasmFeatures::GC)
```
We actually need the GC, though, since we allocate (exception) objects, no?

cfallin replied:
Ah, right, I had been assuming we separately checked that `feature = "gc"` was enabled when the `GC` feature is enabled at the wasmparser level, then leaning on that as the load-bearing thing to ensure we have `feature = "gc"`. Addressed in a076222f88 -- switched to an explicit check under `not(feature = "gc")` to ensure `EXCEPTIONS` is not enabled.

cfallin commented on crates/environ/src/vmoffsets.rs:
```diff
@@ -811,6 +811,18 @@ impl<P: PtrSize> VMOffsets<P> {
         0 * self.pointer_size()
     }
 
+    /// The offset of the `vmctx` field.
+    #[inline]
+    pub fn vmtag_import_vmctx(&self) -> u8 {
+        1 * self.pointer_size()
+    }
+
+    /// The offset of the `index` field.
+    #[inline]
+    pub fn vmtag_import_index(&self) -> u8 {
+        2 * self.pointer_size()
+    }
```
Added!

cfallin commented on crates/cranelift/src/func_environ/gc/enabled.rs:
```diff
@@ -403,6 +403,87 @@ pub fn translate_struct_set(
+        accesses.push((field_layout.offset, field_ty.element_type));
+    }
+
+    let mut result = vec![];
```
Done!

cfallin commented on crates/cranelift/src/func_environ.rs:
```diff
@@ -1644,22 +1710,25 @@ impl<'a, 'func, 'module_env> Call<'a, 'func, 'module_env> {
         (func_addr, callee_vmctx)
     }
 
+    fn caller_vmctx(&self) -> ir::Value {
+        self.builder
+            .func
+            .special_param(ArgumentPurpose::VMContext)
+            .unwrap()
+    }
```
(resolving but feel free to unresolve if you'd like a different name here)

cfallin commented on crates/wasmtime/src/runtime/vm/gc/disabled.rs:
```diff
@@ -13,6 +15,28 @@ pub struct VMGcObjectData {
+    fn from(_e: VMExnRef) -> VMGcRef {
+        unreachable!()
+    }
+}
```
Done!

cfallin commented on crates/wasmtime/src/runtime/vm/throw.rs:
```diff
@@ -0,0 +1,131 @@
+
+                    let (handler_tag_instance, handler_tag_index) = unsafe {
+                        InstanceAndStore::from_vmctx(frame_vmctx, |instance| {
+                            let (instance, store) = instance.unpack_mut();
```
Addressed in 3c48653a25 -- I went with a variant of `StoreOpaque::instance_from_vmctx`, except with the additional safety requirement that the `vmctx` be from that store. This is valid/satisfiable from the stack walk because all Wasm frames in an activation (up to entry from host) will belong to instances in a single store; and is nice not just for perf reasons but because I was having trouble reasoning about the aliasing implications of creating an instance pointer *first*, and reading out its store pointer, then comparing it to see if we return `Some` or not. (It would be perfectly valid for another thread to own some other `Store`; that thread could mutate the instance pointers freely per Rust's aliasing rules; so I think we can't have the instance pointer then do the check, we have to know by invariant that it belongs to us first; but fortunately we do.)

cfallin commented on crates/wasmtime/src/runtime/vm/throw.rs:
```diff
@@ -0,0 +1,131 @@
+///
+/// Must be invoked when Wasm is in the stack and control has
+/// re-entered the runtime.
+pub unsafe fn throw(nogc: &mut AutoAssertNoGc, exnref: VMExnRef) -> ! {
```
Addressed in 1f9f65ad4f6bd24d1fe15120845689f1e11bbd46!

cfallin commented on crates/wasmtime/src/runtime/vm/libcalls.rs:
```diff
@@ -1543,3 +1543,23 @@ fn cont_new(
+        .expect("gc ref should be an exception object");
+    unsafe {
+        store.store_opaque_mut().throw_ref(exnref);
+    }
```
Addressed in 1f9f65ad4f6bd24d1fe15120845689f1e11bbd46!

cfallin commented on src/commands/objdump.rs:
```diff
@@ -65,6 +66,10 @@ pub struct ObjdumpCommand {
     /// Whether or not to show information about stack maps.
     #[arg(long, require_equals = true, value_name = "true|false")]
     stack_maps: Option<Option<bool>>,
+
+    /// Whether or not to show information about exception tables.
+    #[arg(long, require_equals = true, value_name = "true|false")]
+    exception_tables: Option<Option<bool>>,
```
Done!

cfallin commented:
@alexcrichton @fitzgen I've now done the refactor we discussed and also added support for Wasm-to-host, host-to-Wasm, and host-to-host (via `Func::new` -> `Func::call`) exception throws. The remaining bit to address is Pulley support (I suspect this should be relatively straightforward now?) and adding `assert_exception` to our WAST runner then enabling spec-tests. I believe the core should be ready for another pass.

alexcrichton commented on crates/wasmtime/src/runtime/vm/libcalls.rs:
```diff
@@ -1548,3 +1548,25 @@ fn cont_new(
     Ok(Some(AllocationSize(ans.cast::<u8>() as usize)))
 }
+
+#[cfg(feature = "gc")]
+unsafe fn get_instance_id(_store: &mut dyn VMStore, instance: Pin<&mut Instance>) -> u32 {
+    instance.id().as_u32()
+}
+
+#[cfg(feature = "gc")]
+unsafe fn throw_ref(
+    mut store: &mut dyn VMStore,
+    _instance: Pin<&mut Instance>,
+    exnref: u32,
+) -> Result<(), TrapReason> {
```
Could `unsafe` from these two functions be dropped?

cfallin replied:
Done!

alexcrichton commented on crates/wasmtime/src/runtime/vm/libcalls.rs:
```diff
@@ -1548,3 +1548,25 @@ fn cont_new(
+        .into_exnref(&*store.unwrap_gc_store().gc_heap)
+        .expect("gc ref should be an exception object");
+    store.as_store_opaque_mut().set_pending_exception(exnref);
+    Err(TrapReason::User(ExceptionTombstone.into()))
```
Would it be possible to have `TrapReason::Exception` directly instead of, even at the lowest layers, encoding through the indirection of `anyhow::Error`?

alexcrichton commented on tests/all/main.rs:
```diff
@@ -122,6 +123,18 @@ pub(crate) fn gc_store() -> Result<wasmtime::Store<()>> {
     Ok(wasmtime::Store::new(&engine, ()))
 }
 
+pub(crate) fn gc_and_exceptions_store() -> Result<wasmtime::Store<()>> {
+    let _ = env_logger::try_init();
+
+    let mut config = wasmtime::Config::new();
+    config.wasm_function_references(true);
+    config.wasm_gc(true);
+    config.wasm_exceptions(true);
+
+    let engine = wasmtime::Engine::new(&config)?;
+    Ok(wasmtime::Store::new(&engine, ()))
+}
```
Could this be updated to disable `wasm_gc` and `wasm_function_references`, or not explictily enable them, since they shouldn't be required?

cfallin replied:
Done!

cfallin commented on crates/wasmtime/src/runtime/func.rs:
```diff
@@ -1265,7 +1269,21 @@ impl Func {
-        func(caller.sub_caller(), params, results)?;
+        match func(caller.sub_caller(), params, results) {
+            Ok(()) => {}
+            #[cfg(feature = "gc")]
+            Err(e) if e.is::<Rooted<ExnRef>>() => {
+                let exnref = e.downcast::<Rooted<ExnRef>>().unwrap();
+                let exnref = VMGcRef::from_raw_u32(exnref.to_raw(&mut caller.store.0).unwrap())
+                    .unwrap()
+                    .into_exnref_unchecked();
+                caller.store.0.set_pending_exception(exnref);
+                return Err(ExceptionTombstone.into());
+            }
```
I tried to bottom out all of the ways that host functions could be provided and it did seem to all lead to `HostFunc` -- perhaps I missed somewhere. This is unfortunately the best we can do though unless we do have a store in `catch_unwind_and_record_trap`, as far as I can tell. I agree the whole situation is kind of horrific right now...

alexcrichton replied:
IMO this required duplication if we don't have a store when handling traps is motivation to push harder on getting a store available while handling traps. I've done this refactoring https://github.com/bytecodealliance/wasmtime/pull/11441 which I believe means that this code here can be scoped to `HostResult` implementations which can now easily access the store.

cfallin replied:
Resolved in refactor (modulo rooting issues below)!

cfallin commented on crates/unwinder/src/stackwalk.rs:
```diff
@@ -77,6 +80,18 @@ impl Frame {
+    pub unsafe fn read_slot(&self, offset: usize) -> Option<usize> {
+        self.sp
+            .map(|sp| unsafe { *(sp.wrapping_add(offset) as *mut usize) })
+    }
```
Updated!

cfallin commented:
I've refactored on top of #11441 and this is indeed much nicer; I've removed the `From` impls that would otherwise pass through boxed exceptions from hostcalls without converting to a pending exception, so I'm more confident we won't miss any additional hostcall paths now.

The current challenge is dealing with rooting: the factor has placed the point at which we capture the `Rooted<ExnRef>` (refactor to wrap that up in a `wasmtime::Exception` TBD) outside the root handle scope for the hostcall, so any newly allocated exception from the host becomes unrooted. Pulling this string further may require additional refactors to put the LIFO-scope management in common across all hostcalls; just starting to look into this. More thoughts (or a pairing session if that'd be easier) welcome, @alexcrichton...

cfallin commented on crates/wasmtime/src/runtime/vm/traphandlers.rs:
```diff
@@ -659,33 +781,52 @@ impl CallThreadState {
+                });
             }
-        };
-        self.unwind.set(Some((reason, backtrace, coredump)));
```
Done!

cfallin commented on crates/wasmtime/src/runtime/vm/traphandlers.rs:
```diff
@@ -62,29 +65,37 @@ fn lazy_per_thread_init() {
-    tls::with(|info| unsafe { info.unwrap().unwind() })
+pub(super) unsafe fn raise_preexisting_trap(store: &mut dyn crate::vm::VMStore) -> ! {
+    let mut nogc = AutoAssertNoGc::new(store.store_opaque_mut());
+    tls::with(|info| unsafe { info.unwrap().unwind(&mut nogc) })
```
Fixed!

cfallin commented on crates/wasmtime/src/runtime/store.rs:
```diff
@@ -2197,6 +2240,24 @@ at https://bytecodealliance.org/security.
+    /// must be done separately.
+    #[cfg(feature = "gc")]
+    pub(crate) fn set_pending_exception(&mut self, exnref: VMExnRef) {
+        self.pending_exception = Some(exnref.into());
```
Done!

cfallin commented:
Actually, I think there are deeper issues with rooting, exceptions, and `Result` / `?` ergonomics that we didn't really address in the RFC, and I'm finding I'm bumping into these design questions now. The current state of this PR has working-but-for-unrooted-ref-panics in `cargo test --test all -- exceptions`, demonstrating a few examples of the issue.

The basic question is: if we return an exception as a GC object boxed in a `Rooted<ExnRef>` (eventually inside a nicely-packaged `wasmtime::Exception` type) in the `Err` arm of a `Result`, how should we expect this to interact nicely with handle scopes?

The most basic example is a host function that allocates an `ExnRef` and returns `Err(exnref.into())`, and has no other scopes of its own. We can fix up this PR by moving the root scope teardown outside of the point that we root a pending exception on the store, so that can work fine.

But in the general case, with user host code creating nested `RootScope`s, we cannot really expect automatic `?` passing-up-the-`Err` to work well. Consider:

```
fn my_hostcall(...) -> Result<()> {
  let scope = RootScope::new(&mut caller);
  helper(&mut scope)?;
  Ok(())
}

fn helper(ctx: impl AsContextMut<'_>, ...) -> Result<()> {
  return Err(ExnRef::new(...).into());
}
```

This will typecheck fine (`ExnRef::new()` returns a `Rooted<ExnRef>`, we currently take that as an `Error` impl directly, we can wrap it up in `wasmtime::Exception` once I do that); but the exception object will be unrooted once the `?` propagates it past the user's own `RootScope`.

More generally, `Err`-carried types should generally expect to *own* their error information. Our lack of lifetimes to denote valid scope is an explicit ergonomic choice, which is great in general, but hides the fact that we will get a dynamic scope-violation (unrooting) error here.

Should we use a `ManuallyRooted` instead? I think that's much worse -- it's way too easy to leak, since its `Drop` impl doesn't unroot (because it can't, because it doesn't have a mut borrow of the `Store` -- again a totally valid choice that is nice in 99% of cases).

It's somewhat tempting to say "don't do that" to all of this -- it's no different than returning any GC ref type at all. If we take this option, we just fix the exact extent of the root scope when calling out to the host, we root any returned exception when it comes back (concretely: put the `HostResult` conversion *inside* the root scope somehow, probably by pulling the scope out to `catch_and_record_traps`). My broader concern here is that I'm not sure that it has the nice error-propagation properties that we expected. Basically, it's a nonlocal/non-composable footgun: anyone creating a scope anywhere inside complex host logic could cause an exception created on deep to blow up dynamically with a panic.

One alternative is to have an explicit `store.throw(exn)` as part of the public API, and then provide a *public-facing* tombstone type that the user can propagate upward through `Result`s. This does make things a little weird in other ways, because it's implicit state (we could have an API around "cancel pending exception" too if someone wants to "catch" it in host code), but it feels the least footgunny in a lot of others. Basically it's extending what we decided to do internally (root the exception on the store, safely) to user code.

The last option does differ from what we agreed to in the RFC (but the RFC is underspecified with respect to rooting behavior in general, on the other hand) -- so I'd want some consensus here, at least, before building that. Any thoughts?

cfallin commented:
> `wasmtime::ThrownException` so that way the user doesn't have to create the tombstone themselves and can immediately return that error upwards the stack.

Indeed, that was my thought as well. I suppose one could actually also have a signature `throw(...) -> Result<Uninhabited, ThrownException>` so that one could do `store.throw(exn)?;` -- or both, `thrown_exception(...) -> ThrownException` and `throw(...)` -- but at this point I'm bikeshedding a bit. I'll prototype this to see how it feels.

cfallin commented on crates/wasmtime/src/runtime/gc/enabled/exnref.rs:
```diff
@@ -626,12 +625,20 @@ impl ExnRef {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        write!(f, "Wasm exception")
+    }
+}
```
Resolved in latest refactor.

cfallin commented on crates/wasmtime/src/runtime/store.rs:
```diff
@@ -1309,6 +1320,26 @@ impl StoreOpaque {
+            vmctx
+                .byte_sub(mem::size_of::<vm::Instance>())
+                .cast::<vm::Instance>()
+                .as_ref()
```
Done in 45415e4886!

cfallin commented:
New final boss appears: Pulley support will, I think, require refactoring how setjmp/longjmp is emulated. Right now Pulley [special-cases the `raise()` libcall](https://github.com/bytecodealliance/wasmtime/blob/f8177c201e466ace97a3ce00be69adae8627ddd1/crates/wasmtime/src/runtime/vm/interpreter.rs#L237-L240) and does an interpreter-register-level longjmp across all Wasm frames.

I believe what we'll ideally want to do instead, to avoid too much duplication in an alternative `raise()` impl, is to actually invoke the real implementation but then indirect the longjmp through a match on the store's executor: either a real longjmp to the real `jmp_buf` in the `CallThreadState`, or a mutation of the interpreter state and a return. Then similarly `wasmtime_unwinder::resume_to_exception_handler` can have another "host" backend that takes the store, fetches the interpreter state, and updates it.

The alternative is to try to get access to the store inside the Pulley `CallIndirectHost` impl and determine its `UnwindState`, but that seems to pull in too much across abstraction boundaries -- for example, it requires Pulley to take the exception payload off the store's root (i.e., have access to the full `Store`, not just some slice of it).

I'm tempted to try to land this first and defer Pulley support to a separate PR, with some help (probably from @alexcrichton) to determine the best path here -- it's somewhat close (the stackwalk works fine!) but definitely not trivial.

cfallin commented:
I've now rebased (twice -- #11461 created more conflicts -- hopefully this lands soon!) and marked as ready. This PR is stacked on top of #11467. All spec-tests should be passing (will watch CI to make sure it goes green here). Thanks for the feedback so far!

alexcrichton commented on crates/wasmtime/src/config.rs:
```diff
@@ -2234,6 +2244,11 @@ impl Config {
         }
 
+        #[cfg(not(feature = "gc"))]
+        if self.enabled_features.contains(WasmFeatures::EXCEPTIONS) {
```
This should test `features`-the-local-variable since that represents the calculation of wasm features after enabled/disabled/defaults/etc.

Also mind folding the `#[cfg]` into the boolean expression with `!cfg!(feature = "gc")`? I personally prefer to favor `cfg!` where possible since it avoids dead code warnings and such (not that such motivations are applicable here)

cfallin replied:
Done!

alexcrichton commented on crates/wasmtime/src/runtime/code_memory.rs:
```diff
@@ -182,6 +186,16 @@ impl CodeMemory {
+        // compiled artifact is as-produced-by this version of
+        // Wasmtime, and we should always produce a correct exception
+        // table (i.e., we are not expecting untrusted data here).
+        #[cfg(debug_assertions)]
```
Similar to above, could this use `if cfg!(debug_assertions) { ... }` to avoid needing a conditional import on `ExceptionTable`? (this'll show up in CI for release builds I think which aren't enabled by default on PRs)

alexcrichton commented on crates/wasmtime/src/runtime/exception.rs:
```diff
@@ -0,0 +1,52 @@
+/// `Error` requires `Display`.
+impl core::fmt::Display for ThrownException {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        <Self as core::fmt::Debug>::fmt(self, f)
```
Personally I'd be in favor of "thrown wasm exception" or something like that as opposed to the generic `ThrownException` this'll otherwise print

alexcrichton commented on crates/wasmtime/src/runtime/store.rs:
```diff
@@ -1047,6 +1067,105 @@ impl<T> Store<T> {
+        self.inner.pending_exception.is_some()
+    }
+
+    /// Invoke a closure, handling any thrown Wasm exception with the given handler.
+    #[cfg(feature = "gc")]
+    pub fn catch<F, Handler, R>(&mut self, f: F, handler: Handler) -> Result<R>
+    where
+        F: FnOnce(StoreContextMut<'_, T>) -> Result<R>,
+        Handler: FnOnce(StoreContextMut<'_, T>, Rooted<ExnRef>) -> Result<R>,
+    {
+        match f(self.as_context_mut()) {
+            Err(e) if e.is::<ThrownException>() => {
+                debug_assert!(self.has_pending_exception());
+                let mut scope = crate::RootScope::new(self);
+                let exn = scope.as_context_mut().take_pending_exception().unwrap();
+                handler(scope.as_context_mut(), exn)
+            }
+            ret @ Ok(_) | ret @ Err(_) => {
+                debug_assert!(!self.has_pending_exception());
+                ret
+            }
+        }
+    }
```
Given that we need to in theory replicate these methods on `StoreContextMut`, `Store`, `Caller`, and other things that implement `AsContextMut`, and given that in theory we also need to eventually add `catch_async` in addition to `catch`, I'd subjectively say this is probably fine to relegate to examples or documentation but otherwise not live here as a first-class function.

cfallin replied:
Sure, removed for now -- always possible to add later if it turns out to be a real ergonomic need...

alexcrichton commented on crates/wasmtime/src/runtime/store.rs:
```diff
@@ -2264,6 +2424,41 @@ at https://bytecodealliance.org/security.
+
+    #[cfg(feature = "gc")]
+    fn take_pending_exception_rooted(&mut self) -> Option<Rooted<ExnRef>> {
+        self.take_pending_exception().map(|vmexnref| {
```
You can use `?` after `self.take_pending_exception()` to avoid indentation if you'd like

alexcrichton commented on crates/wasmtime/src/runtime/store.rs:
```diff
@@ -2264,6 +2424,41 @@ at https://bytecodealliance.org/security.
+    fn take_pending_exception_rooted(&mut self) -> Option<Rooted<ExnRef>> {
+        self.take_pending_exception().map(|vmexnref| {
+            let mut nogc = AutoAssertNoGc::new(self);
+            let cloned = vmexnref.clone(nogc.require_gc_store_mut().unwrap());
+            ExnRef::from_cloned_gc_ref(&mut nogc, cloned.into())
```
Is `clone` the right operation to call here? I would have expected we just thread through the `VMGcRef` as-is because we're effectively moving ownership into the return value

alexcrichton commented on crates/wasmtime/src/runtime/vm/traphandlers.rs:
```diff
@@ -362,11 +384,32 @@ pub enum TrapReason {
 impl From<Error> for TrapReason {
-    fn from(err: Error) -> Self {
-        TrapReason::User(err)
+    #[cfg(feature = "gc")]
+    fn from(error: Error) -> Self {
+        if error.is::<ThrownException>() {
+            TrapReason::Exception
+        } else {
+            TrapReason::User(error)
+        }
+    }
+
+    #[cfg(not(feature = "gc"))]
+    fn from(error: Error) -> Self {
+        TrapReason::User(error)
     }
```
FWIW this is an example of where in my subjective opinion `#[cfg]` is overkill. For example if we have `ThrownException`, a unit struct, in the public API when the `gc` feature is disabled that seems fine to me and worth it to avoid growing more `#[cfg]`

cfallin replied:
Hmm, I pulled on this thread a bit, and adding a `ThrownException` unit struct (or uninhabitable enum) allows this to clean up, but then starts pulling a bunch of other things out of `gc`-only config as well (`TrapReason`, more unwind code depending on where we want to draw the boundary...) and honestly it seems cleaner to me to take the sharp line of types-and-enum-arms-don't-exist-for-exceptions in a GC-disabled build? Otherwise there are a bunch of dynamically unreachable failure points, which seems awkward in its own way...

alexcrichton commented on crates/wasmtime/src/runtime/vm/traphandlers.rs:
```diff
@@ -387,7 +430,7 @@ pub unsafe fn catch_traps<T, F>(
     old_state: &mut EntryStoreContext,
     mut closure: F,
-) -> Result<(), Box<Trap>>
+) -> Result<()>
```
With the changes here and to the `match` below, could you run the `call.rs` benchmark before/after this PR with the `^sync.*no-hook.*host-to-wasm.* typed - nop$` regex? If there's no real change, yay! If there's a regression, mind spot-checking to see if anything jumps out?

alexcrichton replied:
That's reasonable enough for now IMO. I suspect we might be able to recover that by shuffling around things, maybe adding `#[cold]`, tweaking slow paths, etc. That being said it's still solidly in the realm of "this shouldn't matter and it's still fast", so I think it's fine to defer such theoretical tweaks to a future point in time.

alexcrichton commented on crates/wasmtime/src/runtime/vm/traphandlers.rs:
```diff
@@ -704,11 +868,43 @@ impl CallThreadState {
-        debug_assert!(self.jmp_buf.get() != CallThreadState::JMP_BUF_INTERPRETER_SENTINEL);
-        unsafe {
-            traphandlers::wasmtime_longjmp(self.jmp_buf.get());
+    unsafe fn unwind(&self, store: &mut dyn VMStore) -> ! {
```
Would it be possible to change this to `Option<usize>` representing the payload? That way `UnwindToHost` would assert it's `None` and `UnwindToWasm` would assert it's `Some`.

alexcrichton replied:
Nah that seems reasonable, this is a feeble attempt by me to future-proof Pulley support a bit more. Let's just defer and deal with it in total later.

alexcrichton commented on crates/wast/src/wast.rs:
```diff
@@ -479,6 +479,21 @@ where
+        match result {
+            Outcome::Ok(values) => bail!("expected exception, got {values:?}"),
+            Outcome::Trap(err) if err.is::<ThrownException>() => {
+                // Discard the thrown exception.
+                let _ = self
+                    .store
+                    .take_pending_exception()
+                    .expect("there should be a pending exception on the store");
```
Reading this over, the semantics of `set_pending_exception` are such that it panics if the exception is already set, right? Would that mean that if an embedder forgot to call `take_pending_exception` and then reentered wasm after an exception, would that panic on another exception?

If not, yay! If so, I might say we should instead allow `set_pending_exception` with whatever and it just overwrites whatever was there previously.

cfallin replied:
We did indeed have "panic on re-setting pending exception", due to [your review comment here](https://github.com/bytecodealliance/wasmtime/pull/11326#discussion_r2279374762) suggesting that we `debug_assert` to ensure no prior exception. That said, now that we have the "pending exception slot" API, this is user-exposed and clearly a different tradeoff, so I've removed that assert.

alexcrichton reviewed (commented):
Mind updating a few bits and pieces of documentation as well with this? The ones I can think of are:

* [The wasm proposals table](https://github.com/bytecodealliance/wasmtime/blob/main/docs/stability-wasm-proposals.md#wasm-proposals)
* [The architecture table for supported wasm proposals](https://github.com/bytecodealliance/wasmtime/blob/main/docs/stability-tiers.md#x86_64)

Also, apart from Pulley and fuzzing possibly, are there other known entities to record in issues before landing? E.g. anything like an open question about performance, API, etc.

I should also clarify that I so far haven't looked at the wasmtime-cranelift side of things at all, I figured I'd defer to @fitzgen for that, but Nick if you'd like I'm happy to review it here too.

cfallin commented:
> Mind updating a few bits and pieces of documentation as well with this? The ones I can think of are:
> 
>     * [The wasm proposals table](https://github.com/bytecodealliance/wasmtime/blob/main/docs/stability-wasm-proposals.md#wasm-proposals)
> 
>     * [The architecture table for supported wasm proposals](https://github.com/bytecodealliance/wasmtime/blob/main/docs/stability-tiers.md#x86_64)
> 
> 
> Also, apart from Pulley and fuzzing possibly, are there other known entities to record in issues before landing? E.g. anything like an open question about performance, API, etc.

Docs updated!

The two major remaining bits are Pulley support and ensuring exception support is fuzzed. I'll work on at least the latter, and I'll file issues for both once this merges.

All comments addressed modulo those still open above -- thanks!

cfallin commented on crates/cranelift/src/translate/code_translator.rs:
```diff
@@ -4132,3 +4189,74 @@ fn bitcast_wasm_params(
+    let exn_ref = builder.func.dfg.block_params(block)[0];
+    if needs_stack_map {
+        builder.declare_value_needs_stack_map(exn_ref);
+    }
```
Fixed!

cfallin commented on crates/cranelift/src/func_environ/gc/enabled.rs:
```diff
@@ -409,6 +412,119 @@ pub fn translate_struct_set(
+    tag_index: TagIndex,
+    exn_ref: ir::Value,
+) -> WasmResult<SmallVec<[ir::Value; 4]>> {
+    log::trace!("translate_exn_unbox({tag_index:?} {exn_ref:?})");
```
Done!

cfallin commented on crates/cranelift/src/func_environ/gc/enabled/drc.rs:
```diff
@@ -498,6 +496,83 @@ impl GcCompiler for DrcCompiler {
+        )?;
+        let tag_addr = builder
+            .ins()
+            .iadd_imm(raw_ptr_to_exn, i64::from(tag_offset + 4));
```
Good point -- fixed!

cfallin commented:
@alexcrichton I think this is ready for you again on the three open threads above; and a final LGTM or any other issues you see on the runtime side. Thanks!

alexcrichton reviewed (approved):
All sounds reasonable to me!

I suspect this is likely going to bounce on CI at least once, and for that I'd recommend adding `prtest:full` to the commit messsage to get full CI here on the PR as that should give a much faster cycle time than going through the merge queue. Although that's all moot if it sails through the merge queue on the first try.

alexcrichton commented:
Could you address CI failures in separate commits instead of squashing all back into one? That makes it a bit easier to review after this has merged (and it all gets squashed in the merge queue anyway)

cfallin commented:
Yep, it's almost certainly endianness -- taking a look!

cfallin commented:
Actually, it's an issue with dynamic context reads during the stack-walk, it seems -- s390x ABI is a bit different (stackchains rather than FP-chains) so I suspect this is the issue.

cfallin commented:
s390x turned out to expose a mismatch in the definition of "spillslot offset" for the dynamic context -- the accessor I had exposed returns offset from the fixed storage area, which is ordinarily at `SP+0` *unless* there is an outgoing args area. s390x always has an outgoing args area (per ABI). A win for ISA diversity wrt testing!
